### PR TITLE
HTML API: Implement TEMPLATE and related support

### PR DESCRIPTION
**Pull request by sirreal**
Trac ticket: Core-61576

Adds support for TEMPLATE elements in the HTML Processor.

- **Add some tag handling from #6972**
- **implement most of in_head parsing**
- **Implement step_in_template**

```diff
- Tests: 1527, Assertions: 3166, Skipped: 323.
+ Tests: 1525, Assertions: 3277, Skipped: 233.
```

### CODE REVIEW

**Review by dmsnell — COMMENTED** [review]
@sirreal I've merged `trunk` into this and tried to resolve the conflicts accordingly.

was `step_in_head()` supposed to be in this PR? It's fine if it is, but I didn't know if that was intentional or an artifact of other development.

**Review by dmsnell — APPROVED** [review]
I've given this a full review. Many new tests are passing that were previously skipped, but I had to update the `html5lib` test runner to generate the `content` label and indentation inside `TEMPLATE` elements.

In a few spots I made small adjustments for clarity, e.g. changing `'-' === $tag_name[0]` to `$is_closer`, and replacing `array_push()` with a single element with `$array[] = $single_element`.

On the logic side I added in a check for an empty text node. Perhaps that eventually belongs in `step()` itself, but for now I wanted it in there in case we get all-NULL text nodes.

Additionally I've incorporated as much as I could from `step_in_head()` from #6006. There's a check in the `META` for the document encoding which I didn't think was necessary here, so there are two `if` statements that lack a condition - I'll add that back in once #6006 is ready, unless it merges first.

Please let me know your thoughts @sirreal, also @gziolo or @ockham or @westonruter. Otherwise I'll see if we can't merge this tomorrow (Tuesday, July 30).

**Comment by sirreal** [comment]
> was `step_in_head()` supposed to be in this PR?

TEMPLATE required parsing following the rules for "in head". I knew that there were other PRs implementing that, so I tried to do just enough for TEMPLATE.

File: src/wp-includes/html-api/class-wp-html-processor.php
Comment (sirreal) [inline src/wp-includes/html-api/class-wp-html-processor.php:1209]: Good question. I've read the spec several times and I don't see how. I'm concerned that means we're missing something subtle 😕 

I did check html5ever tree construction, [and it does not seem to do any special adjustment which makes me more confident.](https://github.com/servo/html5ever/blob/fd246b368e4dc9e6ae241051de45dd535af8d76e/html5ever/src/tree_builder/rules.rs#L148-L155)

File: src/wp-includes/html-api/class-wp-html-processor.php
Comment (sirreal) [inline src/wp-includes/html-api/class-wp-html-processor.php:1180]: [In this case, html5ever does account for it, but I still don't see how it behaves differently. ](https://github.com/servo/html5ever/blob/fd246b368e4dc9e6ae241051de45dd535af8d76e/html5ever/src/tree_builder/rules.rs#L128-L138)

I don't think this should block the PR, but I do want to follow-up.

**Comment by sirreal** [comment]
I've done a careful review. This looks good to me, thanks for your additions @dmsnell 👍

File: src/wp-includes/html-api/class-wp-html-processor.php
Comment (westonruter) [inline src/wp-includes/html-api/class-wp-html-processor.php:1127]: Shouldn't this only bail if `$charset` is not UTF-8?

```suggestion
				if ( is_string( $charset ) && ! is_utf8_charset( $charset ) ) {
```

File: src/wp-includes/html-api/class-wp-html-processor.php
Comment (westonruter) [inline src/wp-includes/html-api/class-wp-html-processor.php:1144]: Ditto above. Shouldn't this only bail if the charset is not UTF-8?

I guess this `bail` is essentially a TODO.

**Review by westonruter — APPROVED** [review]
LGTM, with one observation.

Also, the title of this PR should be updated to be more broad, as it is definitely more than just `TEMPLATE` support.

File: src/wp-includes/html-api/class-wp-html-processor.php
Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-processor.php:1127]: yeah there's an interplay here with #6977 that's an artifact due to sharing code but not merging everything at once. the test here is whether we support inferring the character set from a META tag, and that only happens if the parser's charset confidence is `tentative`, but I didn't want to add this code in this PR.

alternatively, we could remove `META` tag support, but I hope that #6977 merges within the next couple of days, making this all moot

**Comment by dmsnell** [comment]
Merged in [[58833]](https://core.trac.wordpress.org/changeset/58833)
https://github.com/wordpress/wordpress-develop/commit/dfd1de09afbb5b3c4c7a959b20a1f5fa7011aa46

**Pull request closed without merging**
